src/prefect/runner/storage.py

```diff
@@ -165,19 +172,47 @@ def _repository_url_with_credentials(self) -> str:
             elif isinstance(v, SecretStr):
                 credentials[k] = v.get_secret_value()
 
-        formatted_credentials = _format_token_from_credentials(
-            urlparse(self._url).netloc, credentials
-        )
-        if url_components.scheme == "https" and formatted_credentials is not None:
+        return _format_token_from_credentials(urlparse(self._url).netloc, credentials)
+
+    @property
+    def _repository_url_with_credentials(self) -> str:
+        if not self._credentials:
+            return self._url
+
+        url_components = urlparse(self._url)
+
+        if url_components.scheme == "https" and self._formatted_credentials is not None:
             updated_components = url_components._replace(
-                netloc=f"{formatted_credentials}@{url_components.netloc}"
+                netloc=f"{self._formatted_credentials}@{url_components.netloc}"
             )
             repository_url = urlunparse(updated_components)
         else:
             repository_url = self._url
 
         return repository_url
 
+    def _git_config(self) -> List[str]:
+        """
+        Build a git configuration to use when running git commands.
+        """
+        config = {}
+
+        # Submodules can be private. The url in .gitmodules
+        # will not include the credentials, we need to
+        # propagate them down here.
+        if self._include_submodules and self._formatted_credentials is not None:
+            url_components = urlparse(self._url)._replace(path="")
+            if url_components.scheme == "https":
+                with_auth = urlunparse(
+                    url_components._replace(
+                        netloc=f"{self._formatted_credentials}@{url_components.netloc}"
+                    )
+                )
+                without_auth = urlunparse(url_components)
```

Comment:
This bit looks almost exactly the same as what's in `_repository_url_with_credentials` except for the removal of the path URL component. Is there a way that we could reuse `_repository_url_with_credentials` here? I'd be ok with updating `_repository_url_with_credentials` to return the parsed URL instead of a `str` if that helps.

Reply:
That's a good point. Refactored slightly, how does this look?

https://github.com/PrefectHQ/prefect/blob/c9ee71de81bae84f029c8615849fac8520625f0f/src/prefect/runner/storage.py#L176-L206